server/interceptor/interceptor.go

```diff
@@ -206,15 +206,21 @@ func (tp *trillianProcessor) After(ctx context.Context, resp interface{}, handle
 		}
 	}
 	if tokens > 0 && len(tp.info.specs) > 0 {
-		// TODO(codingllama): If PutTokens turns out to be unreliable we can still leak tokens. In
-		// this case, we may want to keep tabs on how many tokens we failed to replenish and bundle
-		// them up in the next PutTokens call (possibly as a QuotaManager decorator, or internally
-		// in its impl).
-		err := tp.parent.qm.PutTokens(ctx, tokens, tp.info.specs)
-		if err != nil {
-			glog.Warningf("Failed to replenish %v tokens: %v", tokens, err)
-		}
-		quota.Metrics.IncReturned(tokens, tp.info.specs, err == nil)
+		// Run PutTokens in a separate goroutine and with a separate context.
+		// It shouldn't block RPC completion, nor it should share the RPC's context deadline.
+		go func() {
```

Comment:
Did you consider a single go routine with a channel of token counts to return, and these funcs which need to return tokens pop them down the channel?

Reply:
I didn't think of that. Do you think it has significant advantages compared to what we have here?

Reply:
I think there are two benefits (of undetermined significance 😄 ):
 - this func becomes ~25% smaller, and focusses on what it _does_ do before returning, so I'd argue would be easier to read
 - also it wouldn't create 2x the number of request-serving goroutines.

wdyt?

Reply:
Discussed offline. I think the TL;DR is that neither of us has a strong opinion, although we seem to lean to different directions. Decided to ask a Go expert (@taknira) so we can turn this into a learning opportunity.

Reply:
As discussed in person, I think goroutines are light-weight enough that having lots of little ones here vs one big one further out doesn't really matter too much.  Also, you'd likely end up having multiple goroutines reading off the channel to prevent it blocking anyway, so it's really a question of having them here or there.

If I were writing this code myself from scratch, I'd probably do it the other way, where this just popped the relevant info (num of tokens and the spec) onto a channel, but as it's already implemented like this I don't have a strong enough opinion either way to ask for it to be changed.